doc/en/man/mqhash.1
.\" Do *not* edit this file; it was automatically generated by ej2man
.\" Look for a name.ej file with the same name as this filename
.\"
.\" Process this file with the following (replace filename.1)
.\" preconv < filename.1 | nroff -man -Tutf8
.\"
.\" Last updated 2022-11-12
.\"
.TH MQHASH 1 "August 2006" MQHASH "MaraDNS reference"
.\" We don't want hyphenation (it's too ugly)
.\" We also disable justification when using nroff
.\" Due to the way the -mandoc macro works, this needs to be placed
.\" after the .TH heading
.hy 0
.if n .na
.\"
.\" We need the following stuff so that we can have single quotes
.\" In both groff and other UNIX *roff processors
.if \n(.g .mso www.tmac
.ds aq \(aq
.if !\n(.g .if '\(aq'' .ds aq \'

  
.SH "NAME"
.PP
mqhash - Simple secure password generator 
.SH "SYNOPSIS"
.PP

.B "mqhash [-n #] [-s] [-u] {data to hash}"
.SH "DESCRIPTION"
.PP

.B "mqhash"
is a simple secure password generator. 
.PP
The program uses MaraDNS\(aq secure random number generator as the 
compression function for a secure hash; the output of this secure hash 
can be used as passwords for various locations on the internet. 
.PP
This program solves the problem of either using the same password on 
multiple web sites, or having so many passwords that it is not 
practical to remember them all. 
.SH "USAGE"
.PP
The first step in using 
.B "mqhash"
as a secure password generator is to set up a master secret from which 
all other passwords are generated. It is important to keep this master 
secret secure; such as on a Linux or BSD machine that is always behind 
a firewall and is current with security updates. 
.PP
This secure secret is put in the file ~/.mhash_prefix. It is important 
that this secret is hard to guess; the security of all generated 
passwords is only as secure as the master secret. 
.PP
Once the ~/.mhash_prefix file is set up, mqhash is run thusly:

.nf
mqhash -s {location} 
.fi

Where {location} is a web site, email address, or any other text string 
that describes where a given password is located. It is recommended 
that one uses a consistent style for {location} so that one can 
remember passwords for web sites that one has not visited for a while. 
Mqhash does not impose a style for remembering passwords; it is up to 
the user to create one. 
.PP
.B "mqhash"
will output four potential passwords that have 32 bits of entropy. If 
more entropy is desired in a password, two 32-bit passwords can be 
joined together to generate a 64-bit password. A 32-bit password will 
protect against casual attacks but can be broken by a determined 
attacker with extensive resources attacking a website that does not 
lock out a user after too many failed attempts. A 64-bit password is 
immune to even a very determined attacker. 
.PP

.in -3
\fBOPTIONS\fR
.PP

.TP 4
-n 
It is wise to periodically change ones password on sites that one uses 
frequently. This allows one to continue to have passwords after the 
four initial passwords have already been used; this can have a value 
between 2 and 9. 
.TP 4
-s 
The normal mode for mqhash: To create a secure password based on both 
the contents of ~/.mhash_prefix and the final argument to mqhash. 
.TP 4
-u 
This will generate a cryptographic hash out of the final argument sent 
to mqhash. This is useful when one does not need a secure password, but 
just wants to hash a short string. 
.PP

.SH "LEGAL DISCLAIMER"
.PP
THIS SOFTWARE IS PROVIDED BY THE AUTHORS \(aq\(aqAS IS\(aq\(aq AND ANY 
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS 
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR 
BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
.SH "AUTHORS"
.PP
Sam Trenholme (http://www.samiam.org) is responsible for this man page. 
 



CHANGELOG
MaraDNS changelog

   maradns-3.5.0020:
   This is a stable release of MaraDNS:

     * New parameter: source_ip4, to specify the source IP when sending a
       query to an upstream or authoritative DNS server.
     * Makefile.centos8 file for Deadwood renamed Makefile.Ubuntu2004 (i.e.
       it’s now a Makefile for Ubuntu 20.04)

     (2021-05-??)

   maradns-3.5.0019:
   This is a stable release of MaraDNS:

     * One line change to zoneserver.c to make it work better with systemd
     * Synthetic IP generator example (e.g. 10.1.2.3.ip4.internal 
       resolves to 10.1.2.3) added to coLunacyDNS documentation

     (2021-03-16)

   maradns-3.5.0018:
   This is a stable release of MaraDNS:

     * coLunacyDNS updated to 1.0.010 (Ubuntu 20.04 testing found a
       minor select() bug)
     * Since RedHat has broken their pinky promise to support CentOS 8
       until 2029, MaraDNS has moved from CentOS 8 to Ubuntu 20.04 LTS
       (RPM .spec files removed; Docker container now runs Ubuntu 20.04)

     (2020-12-20)

   maradns-3.5.0017:
   This is a stable release of MaraDNS:

     * coLunacyDNS update: We can now specify the “Authoritative”
       and “Recursion available” flags in the reply.
     * coLunacyDNS update: We can now specify a TTL for the reply,
       to be anywhere from 0 seconds (do not cache) to a little over
       90 days.
     * coLunacyDNS version updated to be 1.0.009.

     (2020-10-02)

   maradns-3.5.0016:
   This is a stable release of MaraDNS:

     * Unstable mmLunacyDNS code removed from tree (coLunacyDNS can do
       anything mmLunacyDNS could do)
     * coLunacyDNS bug fixes: We return with an error if the Lua code
       attempts to return an invalid IPv4 address to the client (before,
       the code incorrectly returned 255.255.255.255)
     * coLunacyDNS returns helpful errors if the processQuery return 
       value is invalid in various ways.
     * coLunacyDNS now has 100%* testing coverage. *Some sanity tests
       which protect coLunacyDNS from security threats which can not
       be readily reproduced are disabled in testing mode.
     * coLunacyDNS is now at version 1.0.008

     (2020-09-01)

   maradns-3.5.0015:
   This is a stable release of MaraDNS:

     * coLunacyDNS is now a stable release (1.0.007) with a full
       SQA testing suite and well over 90% code coverage in its tests.
     * mmLunacyDNS has been removed; coLunacyDNS can do everything
       mmLunacyDNS could do, and this saves me the bother of
       maintaining two code bases.
     * askmara now compiles with IPv6 support (the code has been
       there, but is finally getting enabled)

     (2020-08-29)

   maradns-3.5.0014:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable):

     * coLunacyDNS’s Lua script can now specify IPv6 addresses in
       standard “colon” format, e.g. co1Data="2001:db8::1"
     * coLunacyDNS now handles ANY (and HINFO) queries as per RFC8482
     * coLunacyDNS documentation updates: Various cleanup.  Also,
       the example coLunacyDNS .lua files now return “not there”
       when we ask for a hostname which does not have a given record.

     (2020-08-20)

   maradns-3.5.0013:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable):

     * coLunacyDNS now has support for binding to IPv6 addresses.  Both
       the *NIX (Linux) and the Windows32 binary can bind to an IPv6
       socket.
     * Some other bug fixes and cleanup, mainly with coLunacyDNS.

     (2020-08-19)

   maradns-3.5.0012:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable): 

     * mmLunacyDNS security fix: We now use a secure hash compression
       function (HalfSipHash-1-3) for string hashing.
     * coLunacyDNS: hash compression function updated from 64-bit
       SipHash-2-4 to 32-bit HalfSipHash-1-3.  Compile time warnings
       removed from code.
     * lunacy: The code by default now uses HalfSipHash-1-3 for string
       hash compression.  Default compile optimization is now -O3
 
     (2020-08-12)

   maradns-3.5.0011:
   This is a stable release of MaraDNS:

     * min_ttl parameter added; this is the minimum time we keep a 
       record in the cache (in seconds)
     * Deadwood now compiles with IPv6 support by default.  For
       systems without IPv6 support, -DNOIP6 can be set when
       compiling Deadwood.
     * Automated tests now all run inside of Podman (Docker) container
       and all pass.  Tests are now completely automated, and can run
       from cron (and can be adapted to run inside Jenkins). 

     (2020-08-10)

   maradns-3.5.0010:
   This is a stable release of MaraDNS:

     * Hotfix: coLunacyDNS no longer fails after 20 calls to 
       processQuery() (we now properly clean the main stack before 
       calling processQuery() in a co-routine).
     * Security update: MaraDNS, Deadwood, and Duende now default to
       the user ID 707 instead of 99/66.  This minimizes the chances
       of the user used by MaraDNS being used by other processes, 
       which could be a security leak under some circumstances.  The
       problem with running multiple services as "nobody" is that
       the "nobody" account is only as secure as the least secure
       service running as that account.
     * coLunacyDNS feature update: coLunacyDNS can now open and
       read files (for security reasons, only in the same directory
       coLunacyDNS is running in).  In addition, the code to implement 
       IPv6 sockets is well under way.

     (2020-08-06)

   maradns-3.5.0009:
   This is a stable release of MaraDNS:

     * Add new program: coLunacyDNS.  This is a DNS server which runs a
       Lua function every time it gets a DNS query.  It uses Lua
       threads ("co-routines") to have a function which can get a
       DNS packet from an upstream server and return the result for

tools/misc/mqhash.c
/* Copyright (c) 2002-2012 Sam Trenholme
 *
 * TERMS
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * This software is provided 'as is' with no guarantees of correctness or
 * fitness for purpose.
 */

/* This code is based on the aeshash.pdf file, written by Bram Cohen and
   Ben Laurie.  The code currently uses the 128-bit key and block size of
   MaraDNS's hasher, making a 128-bit hash */

/* Note that you need to change the cipher we call to change these
   constants */
#define HASH_BITS 128
#define HASH_BYTES (HASH_BITS / 8)

#include "../../rng/rng-api-fst.h"
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

/* Routine that runs the compression funciton of our hash.
   Input: Plaintext we wish to compress (which is HASH_BYTES long),
          current state of the hash
   Output: The current state of the hash is modified
           1 on success, -1 on fail
*/

int hash_compress(char *input, char *state) {
    MARA_BYTE r_inBlock[HASH_BYTES + 1], r_outBlock[HASH_BYTES + 1],
         r_binKey[HASH_BYTES + 1];
    keyInstance r_keyInst;
    cipherInstance r_cipherInst;
    int counter;

#ifdef DEBUG
    for(counter = 0; counter < HASH_BYTES; counter++) {
        if(input[counter] >= 32 && input[counter] < 128)
            printf("%c",input[counter] & 0xff);
        else
            printf("");
        }
    printf("\n");
#endif /* DEBUG */

    /* Copy over the input of the hash to the key for the cipher */
    for(counter = 0; counter < HASH_BYTES; counter++) {
        r_binKey[counter] = input[counter];
        }

    /* Copy over the state and make it the "plaintext" of the cipher */
    for(counter = 0; counter < HASH_BYTES; counter++) {
        r_inBlock[counter] = state[counter];
        }

    /* Prepare the encryption */
    if(makeKey(&r_keyInst, DIR_ENCRYPT, HASH_BITS, r_binKey) != 1) {
        return -1;
        }

    if(cipherInit(&r_cipherInst, MODE_ECB, NULL) != 1) {
        return -1;
        }

    /* Perform the encryption */
    if(blockEncrypt(&r_cipherInst, &r_keyInst, r_inBlock, HASH_BITS,
                    r_outBlock) != HASH_BITS) {
        return -1;
        }

    /* XOR the ciphertext with the current hash state */
    for(counter = 0; counter < HASH_BYTES; counter++) {
        r_outBlock[counter] ^= state[counter];
        }

    /* Make the modified ciphertext the new state */
    for(counter = 0; counter < HASH_BYTES; counter++) {
        state[counter] = r_outBlock[counter];
        }

    return 1;
    }

/* The Radio Gatun 32-bit core */
/* This determines the word size we use for this particular Radio Gatun
 * implementation; DWR_WORDSIZE needs to be a multiple of 8.  Note also
 * that dwr_rng() needs to be rewritten if these values are changed. */
#define DWR_WORD uint32_t
#define DWR_WORDSIZE 32

/* These are hard coded in the Radio Gatun specification */
#define DWR_MILLSIZE 19
#define DWR_BELTROWS 3
#define DWR_BELTCOL 13
#define DWR_BELTFEED 12

/* A structure contining a RadioGatun state */
typedef struct {
        DWR_WORD *mill;
        DWR_WORD *belt;
        int32_t index;
} dwr_rg;
/* This is the mill part of the RadioGatun algorithm */
void dwr_mill(DWR_WORD *a) {
        DWR_WORD A[DWR_MILLSIZE];
        DWR_WORD x;
        int i = 0;
        int y = 0;
        int r = 0;
        int z = 0;
        for(i = 0; i < DWR_MILLSIZE ; i++) {
                y = (i * 7) % DWR_MILLSIZE;
                r = ((i * (i + 1)) / 2) % DWR_WORDSIZE;
                x = a[y] ^ (a[ ((y + 1) % DWR_MILLSIZE) ] |
                    (~a[ ((y + 2) % DWR_MILLSIZE) ]));
                A[i] = (x >> r) | (x << (DWR_WORDSIZE - r));
        }
        for(i = 0; i < DWR_MILLSIZE ; i++) {
                y = i;
                r = (i + 1) % DWR_MILLSIZE;
                z = (i + 4) % DWR_MILLSIZE;
                a[i] = A[y] ^ A[r] ^ A[z];
        }
        a[0] ^= 1;
}

/* This is the belt part of the RadioGatun algorithm */
void dwr_belt(DWR_WORD *a, DWR_WORD *b) {
        DWR_WORD q[DWR_BELTROWS];
        int s = 0;
        int i = 0;
        int v = 0;
        for(s = 0; s < DWR_BELTROWS ; s++) {
                q[s] = b[((s * DWR_BELTCOL) + DWR_BELTCOL - 1)];
        }
        for(i = DWR_BELTCOL - 1; i > 0; i--) {
                for(s = 0; s < DWR_BELTROWS ; s++) {
                        v = i - 1;
                        if(v < 0) {
                                v = DWR_BELTCOL - 1;
                        }
                        b[((s * DWR_BELTCOL) + i)] =
                                b[((s * DWR_BELTCOL) + v)];
                }
        }
        for(s = 0; s < DWR_BELTROWS; s++) {
                b[(s * DWR_BELTCOL)] = q[s];
        }
        for(i = 0; i < DWR_BELTFEED ; i++) {
                s = (i + 1) + ((i % DWR_BELTROWS) * DWR_BELTCOL);
                b[s] ^= a[(i + 1)];
        }
        dwr_mill(a);
        for(i = 0; i < DWR_BELTROWS; i++) {
                a[(i + DWR_BELTCOL)] ^= q[i];
        }
}

/* Destroy an already created RadioGatun state */
void dwr_zap(dwr_rg *tozap) {
        if(tozap != 0) {
                if(tozap->mill != 0) {
                        free(tozap->mill);
                }
                if(tozap->belt != 0) {
                        free(tozap->belt);
                }
                free(tozap);
        }
        return;
}

/* Create a new blank RadioGatun state (this is private and only called
 * from dwr_init_rg() ) */
dwr_rg *dwr_new() {
        int q = 0;
        dwr_rg *new = 0;
        DWR_WORD *a = 0, *b = 0;

        new = malloc(sizeof(dwr_rg));
        if(new == 0) {
                goto catch_dwr_new;
        }
        new->mill = malloc((DWR_MILLSIZE + 1) * sizeof(DWR_WORD));
        new->belt = malloc(((DWR_BELTROWS * DWR_BELTCOL) + 1) *
                           sizeof(DWR_WORD));
        new->index = 0;
        a = new->mill;
        b = new->belt;
        if(a == 0 || b == 0) {
                goto catch_dwr_new;
        }
        for(q = 0; q < DWR_MILLSIZE; q++) {
                a[q] = 0;
        }
        for(q = 0; q < DWR_BELTROWS * DWR_BELTCOL; q++) {
                b[q] = 0;
        }
        return new;
catch_dwr_new:
        dwr_zap(new);
        return 0;
}

#define dwr_input_map()         for(c = 0; c < 3; c++) { \
                                        b[c * 13] ^= p[c]; \
                                        a[16 + c] ^= p[c]; \
                                } \
                                dwr_belt(a,b)

/* Create a Radio Gatun state, using the contents of a null-terminated string
 * as the input */
dwr_rg *dwr_init_rg(char *obj) {
        DWR_WORD p[3];
        int q = 0;
        int c = 0;
        int r = 0;
        int done = 0;
        dwr_rg *new = 0;
        DWR_WORD *a = 0, *b = 0;
        int index = 0;
        int32_t counter = 0;

        new = dwr_new();
        if(new == 0) {
                return 0;
        }
        a = new->mill;
        b = new->belt;
        for(counter = 0; counter < 16777218; counter++) {
                p[0] = p[1] = p[2] = 0;
                for(r = 0; r < 3; r++) {
                        for(q = 0; q < DWR_WORDSIZE; q+=8) {
                                int x = 0;
                                x = (int)*(obj + index);
                                index++;
                                x &= 0xff;
                                if(x == 0) {
                                        done = 1;
                                        x = 1; /* Append with single byte
                                                * w/ value of 1 */
                                }
                                p[r] |= x << q;
                                if(done == 1) {
                                        dwr_input_map();
                                        for(c = 0; c < 16; c++) {
                                                dwr_belt(a,b);
                                        }
                                        return new;
                                }
                        }
                }
                dwr_input_map();
        }
        return 0;
}

/* Given a RadioGatun state, generate a psuedo-random number. */
DWR_WORD dwr_rng(dwr_rg *in) {
        DWR_WORD *o;
        DWR_WORD out;
        if(in == 0) {
                return 0;
        }
        if(DWR_WORDSIZE != 32) {
                return 0;
        }
        o = in->mill + 1;
        if(in->index >= 100000000) {
                in->index = 0; /* I am considering rekeying here */
        }
        if(in->index % 2 == 0) {
                dwr_belt(in->mill, in->belt);
                in->index++;
                out = o[0];
        } else {
                in->index++;
                out = o[1];
                }
        /* Endian issues */
        out = ((out & 0xff0000) >> 8) | ((out & 0xff000000) >> 24) |
              ((out & 0xff00) << 8) | ((out & 0xff) << 24);
        return out;
}

/* The main routine.  This reads a file specified on the command line,
   then makes a hash out of that file. */

main(int argc, char **argv) {
    char state[HASH_BYTES + 1], input[HASH_BYTES + 1];

    int readed, counter;
    FILE *sh;
    char copy[HASH_BYTES * 43];
    int n = 0;
    int q = 0;

    unsigned int len = 0;

    /* Check the command line argument */
    if(argc != 3 && argc != 5) {
        if(argc >= 1) {
            printf("Usage: %s [-n #] [-s] [-u] {data to hash}\n",argv[0]);
            exit(1);
            }
        else {
            printf("Usage: <this program> [-n #] [-s] [-u] {data to hash}\n");
            exit(2);
            }
        }

    /* Initialize the state */
    for(counter = 0; counter < HASH_BYTES; counter++) {
        state[counter] = 0xff;
        }

    if(argc == 5 && (*(argv[1]) != '-' || *(argv[1] + 1) != 'n')) {
            printf("Usage: <this program> [-n #] [-s] [-u] {data to hash}\n");
            exit(3);
            }
    else if(argc == 5) {
            n = atoi(argv[2]);
            if(n < 1 || n > 900) {
                printf("n must be between 1 and 900\n");
                exit(4);
                }
            }
    /* Open up what we prepend to the hash if -s is in argv */
    if((argc == 3 && *(argv[1]) == '-' && *(argv[1] + 1) == 's') ||
       (argc == 5 && *(argv[3]) == '-' && *(argv[3] + 1) == 's')) {
            char fp[100];
            int zork;

            if(strncpy(fp,getenv("HOME"),50) == NULL) {
                    perror("Problem copying string");
                    exit(35);
            }
            if(strcat(fp,"/.mhash_prefix") == NULL) {
                    perror("Problem making string");
                    exit(36);
            }
            if((sh = fopen(fp,"rb")) == NULL) {
                perror("Could not open file ~/.mhash_prefix");
                exit(5);
                }
            /* Get only one line from this file: A string we put at
             * the beginning of the hash we will make */
            for(counter = 0; counter < 85; counter++) {
                copy[counter] = 0;
                }
            fgets(copy,79,sh);
            fclose(sh);
            counter = strnlen(copy,85);
            counter--;
            /* Remove :, which is a metacharacter */
            for(zork = 0; zork <= counter; zork++) {
                if(copy[zork] == ':') {
                    copy[zork] = '@';
                    }
                }
            copy[counter] = ':';
            counter++;
            }
    else if((argc == 3 && *(argv[1]) == '-' && *(argv[1] + 1) == 'q') ||
       (argc == 5 && *(argv[3]) == '-' && *(argv[3] + 1) == 'q')) {
            char fp[100];
            int zork;

            if(strncpy(fp,getenv("HOME"),50) == NULL) {
                    perror("Problem copying string");
                    exit(35);
            }
            if(strcat(fp,"/.mhash_prefix2") == NULL) {
                    perror("Problem making string");
                    exit(36);
            }
            if((sh = fopen(fp,"rb")) == NULL) {
                perror("Could not open file ~/.mhash_prefix2");
                exit(5);
                }
            /* Get only one line from this file: A string we put at
             * the beginning of the hash we will make */
            for(counter = 0; counter < 85; counter++) {
                copy[counter] = 0;
                }
            fgets(copy,79,sh);
            fclose(sh);
            counter = strnlen(copy,85);
            counter--;
            /* Remove :, which is a metacharacter */
            for(zork = 0; zork <= counter; zork++) {
                if(copy[zork] == ':') {
                    copy[zork] = '@';
                    }
                }
            copy[counter] = ':';
            counter++;
            q = 1;
            }
    else if((argc == 3 && *(argv[1]) == '-' && *(argv[1] + 1) == 'u') ||
       (argc == 5 && *(argv[3]) == '-' && *(argv[3] + 1) == 'u')) {
       counter = 0;
       }
    else {
       printf("Usage: <this program> [-n #] [-s] [-u] {data to hash}\n");
       exit(6);
       }

    if(strnlen(argv[argc - 1],HASH_BYTES * 35) >= HASH_BYTES * 31) {
         printf("Hash input is too long!\n");
         }

    readed = counter;
    for(;counter < HASH_BYTES * 42; counter++)
        copy[counter] = 0;
    counter = readed;
    for(;counter < HASH_BYTES * 41; counter++) {
        /* The ':' is always a metacharacter */
        if(argv[argc - 1][counter - readed] == ':')
            argv[argc - 1][counter - readed] = '@';
        if(argv[argc - 1][counter - readed] == '\0')
            break;
        copy[counter] = argv[argc - 1][counter - readed];
        }

    /* Initialize the state */
    for(counter = 0; counter < HASH_BYTES; counter++) {
        state[counter] = 0xff;
        }

    for(counter = 0;counter < 128;counter+=16) {
        for(readed = 0; readed < 16; readed++) {
            input[readed] = copy[counter + readed];
            if(input[readed] == '\0')
                break;
            }
        if(input[readed] == '\0' && readed < 16)
            break;
        hash_compress(input,state);
        }

    len = readed + counter;
#ifdef DEBUG
    printf("%d\n",len);
#endif

    /* Pad the final block */
#ifdef ONE_PAD
    if(readed == 0) {
        readed = 1;
        input[0] = 1;
        }
    else if(readed <= (HASH_BYTES / 2) - 1) {
        input[readed] = 1;
        readed++;
        }
    else if(readed < HASH_BYTES) {
        input[readed] = 1;
        readed++;
#else
    if(readed >= HASH_BYTES / 2) {
#endif
        while(readed < HASH_BYTES) {
            input[readed] = 0;
            readed++;
            }
        hash_compress(input,state);
        readed = 0;
        }
    else if(readed == HASH_BYTES) {
        hash_compress(input,state);
#ifdef ONE_PAD
        readed = 1;
        input[0] = 1;
#else
        readed = 0;
#endif
        }

    while(readed < HASH_BYTES - 4) {
        input[readed] = 0;
        readed++;
        }

    /* We count bits, not bytes (as specified in the aeshash pdf document),
       so multiply len times eight */
    len *= 8;

    /* Pad the length to the hash input */
    for(counter = 3; counter >= 0; counter--) {
        input[readed] = (len >> (counter * 8)) & 0xff;
        readed++;
        }
    hash_compress(input,state);
    hash_compress(state,state);

    if(n == 0) {
        printf("marahash: ");

        for(counter = 0; counter < HASH_BYTES; counter++) {
            printf("%02x",state[counter] & 0xff);
            if(counter % 4 == 3) {printf(" ");}
            }
        printf("\n");
        }

    /* OK, now give them the rg32 hash of the same input */
    dwr_rg *rg32;
    rg32=dwr_init_rg(copy);
    if(n != 0) {
        printf("rg32hash: ");
        for(counter = 1; counter < n; counter++) {
            dwr_rng(rg32);
        }
        if(q == 1) {
                for(counter = 1; counter < n; counter++) {
                    dwr_rng(rg32);
                }
                printf("%08x",dwr_rng(rg32));
        }
        printf("%08x ",dwr_rng(rg32));
        printf("%d",n);
        printf("\n");
        }

    }

